cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1128,6 +1129,14 @@ class ObjectInputFile final : public io::RandomAccessFile {
     return bytes_read;
   }
 
+  std::vector<Future<std::shared_ptr<Buffer>>> ReadManyAsync(
+      const io::IOContext& ctx, const std::vector<io::ReadRange>& ranges) override {
+    return RandomAccessFile::ReadManyAsync(
+        ctx, io::internal::CoalesceReadRanges(std::move(ranges),
+                                              coalesceOptions.hole_size_limit,
+                                              coalesceOptions.range_size_limit));
```

Comment:
Since coalescing may merge ranges, you'll need to post-process things so that the returned futures map 1:1 with the original ranges. (May need to refactor things so you don't have to do the coalescing step twice?)

Reply:
Got it, thanks David, I'm writing a test for verifying these cases.